From 174a95e1d24858caafaaa5044f0707d05f8cb95d Mon Sep 17 00:00:00 2001
From: Colin Watson <cjwatson@canonical.com>
Date: Mon, 24 Jan 2011 14:14:10 +0000
Subject: [PATCH 1/3] aptcc: sanitize file descriptor handling

Make sure that apt always has a stdout fd.  Running apt with no fd 1 is
invalid, and causes any maintainer script that writes to stdout (not an
uncommon practice) to die with EBADF.

Close pipes we don't need before running apt.
---
 backends/aptcc/apt.cpp |    9 ++++-----
 1 files changed, 4 insertions(+), 5 deletions(-)

diff --git a/backends/aptcc/apt.cpp b/backends/aptcc/apt.cpp
index 55c9a62..78935a1 100644
--- a/backends/aptcc/apt.cpp
+++ b/backends/aptcc/apt.cpp
@@ -1783,9 +1783,11 @@ cout << "How odd.. The sizes didn't match, email apt@packages.debian.org";
 			close(writeToChildFD[0]);
 			_exit(1);
 		}
+		close(writeToChildFD[0]);
 
-		// close Forked stdout and the read end of the pipe
-		close(1);
+		// close pipes we don't need
+		close(readFromChildFD[0]);
+		close(writeToChildFD[1]);
 
 		// Change the locale to not get libapt localization
 		setlocale(LC_ALL, "C");
@@ -1814,10 +1816,7 @@ cout << "How odd.. The sizes didn't match, email apt@packages.debian.org";
 		// dump errors into cerr (pass it to the parent process)
 		_error->DumpErrors();
 
-		close(readFromChildFD[0]);
-		close(writeToChildFD[1]);
 		close(readFromChildFD[1]);
-		close(writeToChildFD[0]);
 
 		_exit(res);
 	}
-- 
1.7.2.3

